From 3bbbb0204a4bcdc3b6e03d6f3e5c93426d6178b9 Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@nxp.com>
Date: Thu, 12 Apr 2018 14:34:36 +0800
Subject: [PATCH 3700/5242] MLK-17631-3 usb: cdns3: gadget: fix NULL pointer
 deference issue for usbtest

commit  bf9a985249442335568a7e4659bec45e27259f72 from
https://source.codeaurora.org/external/imx/linux-imx.git

When running usbtest case 9, it meets below oops, the reason for this
issue is the EP2-IN is not enabled, but the related interrupt for it
occurs, the structure usb_request for it is NULL. The software should
make sure there is no oops even the hardware triggers wrong interrupt.

[  476.636715]  gadget-cdns3: EP_STS: 00000804
[  476.636726] Unable to handle kernel NULL pointer dereference at virtual address 00000002
[  476.636729] pgd = ffff0000095f1000
[  476.636736] [00000002] *pgd=00000008bfffe003, *pud=00000008bfffd003, *pmd=0000000000000000
[  476.636741] Internal error: Oops: 96000006 [#1] PREEMPT SMP
[  476.636750] Modules linked in: g_zero vivante ipv6
[  476.636757] CPU: 0 PID: 0 Comm: swapper/0 Not tainted 4.9.88-04771-g7b3994b #1669
[  476.636759] Hardware name: Freescale i.MX8QXP MEK (DT)
[  476.636762] task: ffff0000093c0780 task.stack: ffff0000093b0000
[  476.636774] PC is at single_td_ep_dequeue+0x50/0x180
[  476.636780] LR is at cdns_irq_handler_thread+0x21c/0x6c0
[  476.636783] pc : [<ffff0000088b6aa8>] lr : [<ffff0000088b6214>] pstate: 600001c5
[  476.636785] sp : ffff80083ff40e40
[  476.636790] x29: ffff80083ff40e40 x28: ffff80083ae0c418
[  476.636795] x27: 0000000000040000 x26: 0000000000000804
[  476.636800] x25: ffff0000095624f8 x24: ffff80083b1047a0
[  476.636804] x23: ffff000009187718 x22: ffff0000095619b8
[  476.636809] x21: ffff80083b104000 x20: ffff80083ae0c418
[  476.636813] x19: 0000000000000000 x18: 0000000000000010
[  476.636818] x17: 0000ffff968c5e30 x16: ffff000008b14998
[  476.636823] x15: 0000000000000008 x14: 00000000fffffff0
[  476.636827] x13: ffff0000095a0068 x12: ffff0000093ce600
[  476.636832] x11: ffff0000093ce000 x10: ffff00000959d6d8
[  476.636837] x9 : 0000000000000001 x8 : ffff80083c8eb43f
[  476.636841] x7 : 0000000000000000 x6 : 000000000811c50d
[  476.636845] x5 : 00ffffffffffffff x4 : 0000000000000000
[  476.636850] x3 : 00000000000001c0 x2 : 0000000000000000
[  476.636854] x1 : 0000000000000000 x0 : ffff80083a607410
[  476.636855]
[  476.636859] Process swapper/0 (pid: 0, stack limit = 0xffff0000093b0020)
[  476.636862] Stack: (0xffff80083ff40e40 to 0xffff0000093b4000)
[  476.636864] Call trace:
[  476.636869] Exception stack(0xffff80083ff40c70 to 0xffff80083ff40da0)
[  476.636874] 0c60:                                   0000000000000000 0000ffffffffffff
[  476.636879] 0c80: ffff80083ff40e40 ffff0000088b6aa8 ffff80083ff40db0 ffff80083ff40db0
[  476.636885] 0ca0: ffff80083ff40d80 00000000ffffffd8 ffff80083bb17500 0000000000000000
[  476.636889] 0cc0: ffff8008344f6a00 0000000000000000 0000000000000000 0000000000000000
[  476.636895] 0ce0: ffff80083ff40cf0 ffff0000088f8b9c ffff80083ff40d20 ffff0000088f9270
[  476.636900] 0d00: ffff80083bb17500 0000000000000000 ffff80083a607410 0000000000000000
[  476.636905] 0d20: 0000000000000000 00000000000001c0 0000000000000000 00ffffffffffffff
[  476.636910] 0d40: 000000000811c50d 0000000000000000 ffff80083c8eb43f 0000000000000001
[  476.636916] 0d60: ffff00000959d6d8 ffff0000093ce000 ffff0000093ce600 ffff0000095a0068
[  476.636921] 0d80: 00000000fffffff0 0000000000000008 ffff000008b14998 0000ffff968c5e30
[  476.636927] [<ffff0000088b6aa8>] single_td_ep_dequeue+0x50/0x180
[  476.636932] [<ffff0000088b6214>] cdns_irq_handler_thread+0x21c/0x6c0
[  476.636937] [<ffff0000088b1f0c>] cdns3_irq+0x44/0x98
[  476.636945] [<ffff00000810b0cc>] __handle_irq_event_percpu+0x9c/0x128
[  476.636950] [<ffff00000810b174>] handle_irq_event_percpu+0x1c/0x58
[  476.636955] [<ffff00000810b1f8>] handle_irq_event+0x48/0x78
[  476.636961] [<ffff00000810eb90>] handle_fasteoi_irq+0xb8/0x1b0
[  476.636967] [<ffff00000810a1b4>] generic_handle_irq+0x24/0x38
[  476.636972] [<ffff00000810a81c>] __handle_domain_irq+0x5c/0xb8
[  476.636978] [<ffff00000808180c>] gic_handle_irq+0xbc/0x168
[  476.636981] Exception stack(0xffff0000093b3de0 to 0xffff0000093b3f10)
[  476.636987] 3de0: 0000000000000000 0000000000000000 0000000000000001 0000000000000000
[  476.636992] 3e00: 00000000000001c0 0100000000000000 00000000eb3aea81 ffff80083ff47b38
[  476.636998] 3e20: ffff0000093c1060 ffff0000093b0000 0000000000000880 000000001469197d
[  476.637003] 3e40: 0000000000000018 0000000059362c44 000c28cb00000000 0000554bf0000000
[  476.637008] 3e60: ffff000008b14998 0000ffff968c5e30 0000000000000000 ffff0000093b72e8
[  476.637013] 3e80: ffff0000093b7258 0000000000000001 0000000000000000 0000000000000000
[  476.637019] 3ea0: ffff0000093b0000 ffff00000936da60 ffff000009506fd4 ffff0000093b7000
[  476.637024] 3ec0: ffff0000093b7000 ffff0000093b3f10 ffff00000808599c ffff0000093b3f10
[  476.637029] 3ee0: ffff0000080859a0 0000000000000145 ffff000008116ec4 0000000040000145
[  476.637033] 3f00: ffffffffffffffff 7fffffffffffffff
[  476.637038] [<ffff000008082eec>] el1_irq+0xac/0x120
[  476.637043] [<ffff0000080859a0>] arch_cpu_idle+0x10/0x18
[  476.637050] [<ffff0000080ff478>] cpu_startup_entry+0x118/0x1d8
[  476.637058] [<ffff000008cd54a0>] rest_init+0x80/0x90
[  476.637067] [<ffff000009290b40>] start_kernel+0x378/0x38c
[  476.637072] [<ffff0000092901d8>] __primary_switched+0x5c/0x64
[  476.637078] Code: d1010273 f9401e82 aa1303e1 f943dea0 (39400842)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
[  476.637088] ---[ end trace 21e06658d7988711 ]---
[  476.637091] Kernel panic - not syncing: Fatal exception in interrupt
[  476.637094] SMP: stopping secondary CPUs
[  476.643852] Kernel Offset: disabled
[  476.643854] Memory Limit: none
[  477.339728] ---[ end Kernel panic - not syncing: Fatal exception in interrupt

Reviewed-by: Jun Li <jun.li@nxp.com>
Signed-off-by: Peter Chen <peter.chen@nxp.com>
---
 drivers/usb/cdns3/gadget.c |    3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/usb/cdns3/gadget.c b/drivers/usb/cdns3/gadget.c
index 7144d70..14ab65f 100644
--- a/drivers/usb/cdns3/gadget.c
+++ b/drivers/usb/cdns3/gadget.c
@@ -928,6 +928,9 @@ static int cdns_check_ep_interrupt_proceed(struct usb_ss_endpoint *usb_ss_ep)
 
 		/* get just completed request */
 		request = next_request(&usb_ss_ep->request_list);
+		if (!request)
+			return 0;
+
 		usb_gadget_unmap_request_by_dev(usb_ss->sysdev, request,
 			usb_ss_ep->endpoint.desc->bEndpointAddress
 			& ENDPOINT_DIR_MASK);
-- 
1.7.9.5

